File: internal/sync/containers.go
package sync

import (
	"context"
	"encoding/json"

	"github.com/iskry/cask/internal/config"
	"github.com/iskry/cask/internal/executor"
)

// ContainerSyncManager synchronizes Podman containers between host and config.
type ContainerSyncManager struct {
	Config *config.PodmanConfig
}

func (m *ContainerSyncManager) ResourceType() string { return "containers" }

func (m *ContainerSyncManager) GetHostResources(ctx context.Context, exec executor.Executor) ([]Resource, error) {
	var containers []Resource

	// User-scoped containers
	r, err := exec.Execute(ctx, []string{"podman", "ps", "-a", "--format", "json"})
	if err == nil && r.Success && r.Stdout != "" {
		containers = append(containers, parseContainerJSON(r.Stdout, "user")...)
	}

	// System-scoped containers
	r, err = exec.ExecuteSudo(ctx, []string{"podman", "ps", "-a", "--format", "json"})
	if err == nil && r.Success && r.Stdout != "" {
		containers = append(containers, parseContainerJSON(r.Stdout, "system")...)
	}

	return containers, nil
}

func (m *ContainerSyncManager) GetConfigResources() []Resource {
	if m.Config == nil {
		return nil
	}
	var resources []Resource
	for _, c := range m.Config.Containers {
		resources = append(resources, Resource{
			"name":  c.Name,
			"image": c.Image,
			"scope": string(c.Scope),
		})
	}
	return resources
}

func (m *ContainerSyncManager) ResourceID(r Resource) string {
	name, _ := r["name"].(string)
	return name
}

func (m *ContainerSyncManager) Apply(ctx context.Context, exec executor.Executor, r Resource) bool {
	name, _ := r["name"].(string)
	image, _ := r["image"].(string)
	scope, _ := r["scope"].(string)

	cmd := []string{"podman", "create", "--name", name, image}
	var result *executor.CommandResult
	var err error
	if scope == "system" {
		result, err = exec.ExecuteSudo(ctx, cmd)
	} else {
		result, err = exec.Execute(ctx, cmd)
	}
	return err == nil && result.Success
}

func (m *ContainerSyncManager) Remove(ctx context.Context, exec executor.Executor, id string) bool {
	exec.Execute(ctx, []string{"podman", "stop", id})
	r, err := exec.Execute(ctx, []string{"podman", "rm", "-f", id})
	return err == nil && r.Success
}

func (m *ContainerSyncManager) NeedsUpdate(host, cfg Resource) bool {
	hostImage, _ := host["image"].(string)
	cfgImage, _ := cfg["image"].(string)
	return hostImage != cfgImage
}

func parseContainerJSON(jsonStr string, scope string) []Resource {
	var raw []map[string]any
	if err := json.Unmarshal([]byte(jsonStr), &raw); err != nil {
		return nil
	}
	var resources []Resource
	for _, c := range raw {
		name := ""
		if names, ok := c["Names"].([]any); ok && len(names) > 0 {
			name, _ = names[0].(string)
		} else if n, ok := c["Name"].(string); ok {
			name = n
		}
		image, _ := c["Image"].(string)
		resources = append(resources, Resource{
			"name":  name,
			"image": image,
			"scope": scope,
		})
	}
	return resources
}

File: src/cask/sync/containers.py
"""Podman container sync implementation."""
from __future__ import annotations

from dataclasses import dataclass, field

from cask.config.models import PodmanConfig, ContainerConfig
from cask.executor.protocol import Executor
from cask.managers.podman import PodmanManager
from cask.result import Result


@dataclass
class ContainerResource:
    name: str
    config: ContainerConfig | None = None
    image: str = ""
    ports: list[str] = field(default_factory=list)
    volumes: list[str] = field(default_factory=list)
    read_only: bool = False


class ContainerSync:
    def __init__(self) -> None:
        self._mgr = PodmanManager()

    async def get_host_resources(self, exec: Executor) -> list[ContainerResource]:
        containers = await self._mgr.list_containers(exec)
        resources = []
        for name, info in containers.items():
            if isinstance(info, dict):
                image = info.get("image", "")
                ports = info.get("ports", [])
                volumes = info.get("volumes", [])
            else:
                # Legacy: info is just the image string
                image = info
                ports = []
                volumes = []
            resources.append(ContainerResource(name=name, image=image, ports=ports, volumes=volumes))
        return resources

    def get_config_resources(self, config: PodmanConfig) -> list[ContainerResource]:
        return [
            ContainerResource(
                name=n,
                config=c,
                image=c.image,
                ports=list(c.ports),
                volumes=list(c.volumes),
                read_only=c.security.read_only,
            )
            for n, c in config.containers.items()
        ]

    async def apply(self, resource: ContainerResource, exec: Executor) -> Result:
        if resource.config:
            return await self._mgr.create(resource.name, resource.config, exec)
        return Result(ok=False, message=f"No config for {resource.name}", actions=[])

    async def remove(self, resource_id: str, exec: Executor) -> Result:
        return await self._mgr.remove(resource_id, exec)

    def needs_update(self, host: ContainerResource, config: ContainerResource) -> bool:
        return (
            host.image != config.image
            or sorted(host.ports) != sorted(config.ports)
            or sorted(host.volumes) != sorted(config.volumes)
        )

    def resource_id(self, resource: ContainerResource) -> str:
        return resource.name
